network_plugin/public_nat.py

```diff
@@ -0,0 +1,107 @@
+from cloudify import ctx
+from cloudify import exceptions as cfy_exc
+from cloudify.decorators import operation
+from vcloud_plugin_common import with_vca_client, get_vcloud_config
+from network_plugin import check_ip, save_gateway_configuration, get_vm_ip
+
+VCLOUD_NETWORK_NAME = 'vcloud_network_name'
+PUBLIC_IP = 'public_ip'
+CREATE = 1
+DELETE = 2
+
+
+@operation
+@with_vca_client
+def connect_nat_to_network(vca_client, **kwargs):
+    prepare_network_operation(vca_client, CREATE)
+
+
+@operation
+@with_vca_client
+def disconnect_nat_from_network(vca_client, **kwargs):
+    prepare_network_operation(vca_client, DELETE)
+
+
+@operation
+@with_vca_client
+def connect_nat_to_vm(vca_client, **kwargs):
+    prepare_vm_operation(vca_client, CREATE)
+
+
+@operation
+@with_vca_client
+def disconnect_nat_from_vm(vca_client, **kwargs):
+    prepare_vm_operation(vca_client, DELETE)
+
+
+def prepare_network_operation(vca_client, operation):
+    try:
+        network_name = ctx.source.node.properties['resource_id']
```

Comment:
resource_id doesn't always set to network name. Please see http://getcloudify.org/guide/3.1/plugin-openstack.html#using-existing-resources for external resources usage description. Correct way is to use ctx.source.instance.runtime_properties[VCLOUD_NETWORK_NAME]